File: leetcode.com/1641_count_sorted_vowel_strings/solution.py
class Solution:
    def countVowelStrings(self, n: int) -> int:
        vowels = ('a', 'e', 'i', 'o', 'u')

        @functools.cache
        def dfs(i, r):
            if r == 0:
                return 1

            total = 0

            for j in range(i, len(vowels)):
                total += dfs(j, r - 1)
            return total

        return dfs(0, n)

File: leetcode.com/1641_count_sorted_vowel_strings/solution.c
char vowels[] = {'a', 'e', 'i', 'o', 'u'};

int cache[5][61];

int get(int i, int r)
{
    if (r == 0)
        return 1;

    if (cache[i][r] != -1)
        return cache[i][r];

    int total = 0;

    for (int j = i ; j < 5 ; j++) {
        total += get(j, r - 1);
    }

    cache[i][r] = total;
    return total;
}

int countVowelStrings(int n)
{
    memset(cache, -1, sizeof(cache));
    return get(0, n);
}
